x-pack/plugins/cases/server/saved_object_types/migrations/user_actions.ts

```diff
@@ -58,47 +51,45 @@ export function userActionsConnectorIdMigration(
   }
 }
 
-function isConnectorUserAction(action?: string, actionFields?: string[]): boolean {
-  return (
-    isCreateConnector(action, actionFields) ||
-    isUpdateConnector(action, actionFields) ||
-    isPush(action, actionFields)
-  );
-}
-
-function formatDocumentWithConnectorReferences(
-  doc: SavedObjectUnsanitizedDoc<UserActionUnmigratedConnectorDocument>
+export function payloadMigration(
+  doc: SavedObjectUnsanitizedDoc<UserActions>,
+  context: SavedObjectMigrationContext
 ): SavedObjectSanitizedDoc<unknown> {
-  const { new_value, old_value, action, action_field, ...restAttributes } = doc.attributes;
-  const { references = [] } = doc;
-
-  const { transformedActionDetails: transformedNewValue, references: newValueConnectorRefs } =
-    extractConnectorIdFromJson({
-      action,
-      actionFields: action_field,
-      actionDetails: new_value,
-      fieldType: UserActionFieldType.New,
-    });
+  const originalDocWithReferences = { ...doc, references: doc.references ?? [] };
+  const owner = originalDocWithReferences.attributes.owner;
+
+  try {
+    const { new_value, old_value, action_field, action_at, action_by, action, ...restAttributes } =
+      originalDocWithReferences.attributes;
+
+    const type = getUserActionType(action_field, action);
+    const payload = getPayload(type, action_field, new_value, old_value, owner);
+    const references = removeOldReferences(doc.references);
+    const newAction = action === 'push-to-service' ? Actions.push_to_service : action;
 
-  const { transformedActionDetails: transformedOldValue, references: oldValueConnectorRefs } =
-    extractConnectorIdFromJson({
-      action,
-      actionFields: action_field,
-      actionDetails: old_value,
-      fieldType: UserActionFieldType.Old,
+    return {
+      ...originalDocWithReferences,
+      attributes: {
+        ...restAttributes,
+        action: newAction,
+        created_at: action_at,
+        created_by: action_by,
+        payload,
+        type,
+      },
+      references,
+    };
+  } catch (error) {
+    logError({
+      id: doc.id,
+      context,
+      error,
+      docType: 'user action connector',
+      docKey: 'userAction',
     });
 
-  return {
-    ...doc,
-    attributes: {
-      ...restAttributes,
-      action,
-      action_field,
-      new_value: transformedNewValue,
-      old_value: transformedOldValue,
-    },
-    references: [...references, ...newValueConnectorRefs, ...oldValueConnectorRefs],
-  };
+    return originalDocWithReferences;
```

Comment:
So I ran into this testing the alert schema migration. The mappings for saved objects is actually `dynamic: strict` so because we're changing the saved object schema if we return the original document it'll still crash because when we attempt to insert the a document that doesn't match our mapping. I don't think there's anything else we can do here though to be honest 🤷‍♂️  

I suppose we could try to create as much of a user action as possible? Or fail and just make sure we have a lot of tests to make sure we handle edge cases for our migrations.

Reply:
I decided to create as much of a user action as possible and leave the payload an empty object. This should be a very rare case. Our type guards, unit & integrations test should catch it.

```
return {
      ...originalDocWithReferences,
      attributes: {
        ...restAttributes,
        action: newAction,
        created_at: action_at,
        created_by: action_by,
        payload: {},
        type,
      },
      references: doc.references ?? [],
    };
```